### PR TITLE
IOMCU: added ChibiOS based IO firmware

### PR DESCRIPTION
**Pull request by tridge**
This is an implementation of a F100 based IOMCU firmware using ChibiOS. This PR does not replace the IO firmware we use yet, as it is not yet feature complete, but it is working for basic bench testing.
Most of the work was done by @bugobliterator, with recent additions by me

Note: we should add a "is IO healthy" arming check, and that should check if arming switch change is pending

Remaining issues as of Sep 5th:

- [ ] soft reboot failure, IOMCU fails to connect
- [x] we need an option to use the new IO firmware in builds
- [x] we need a "./waf iofirmware" target
- [ ] needs a lot of testing
- [ ] channels 5 to 8 not functional

### CODE REVIEW

File: Tools/ardupilotwaf/boards.py [hunk 1]
```diff
@@ -338,21 +338,18 @@ def configure_env(self, cfg, env):
...
         ]
 
         if sys.platform == 'cygwin':
             env.CXXFLAGS += ['-DCYGWIN_BUILD']
 
         bldnode = cfg.bldnode.make_node(self.name)
         env.BUILDROOT = bldnode.make_node('').abspath()
-
         env.LINKFLAGS = [
-            '-mcpu=cortex-m4',
```
Comment (khancyr) [inline Tools/ardupilotwaf/boards.py]: I don't think we should lost those flag for F4 cpu

File: libraries/AP_HAL_ChibiOS/UARTDriver_Generic.cpp [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+        return;
+#endif
+    }
+    if(_is_usb) {
+        ((GPIO *)hal.gpio)->set_usb_connected();
+    }
+    _in_timer = true;
+
+    // write any pending bytes
+    n = _writebuf.available();
```
Comment (khancyr) [inline libraries/AP_HAL_ChibiOS/UARTDriver_Generic.cpp]: uint32_t n

Comment (khancyr) [reply libraries/AP_HAL_ChibiOS/UARTDriver_Generic.cpp]: instantiate ret here too

File: libraries/AP_HAL_ChibiOS/UARTDriver_Generic.cpp [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+    if (_is_usb) {
+#if HAL_USE_SERIAL_USB == TRUE
+        if (((SerialUSBDriver*)_serial)->config->usbp->state != USB_ACTIVE) {
+            return;
+        }
+#else
+        return;
+#endif
+    }
+    if(_is_usb) {
```
Comment (khancyr) [inline libraries/AP_HAL_ChibiOS/UARTDriver_Generic.cpp]: if (

File: libraries/AP_HAL_ChibiOS/UARTDriver_Generic.cpp [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+    }
+    size_t ret = _writebuf.write(&c, 1);
+    chMtxUnlock(&_write_mutex);
+    return ret;
+}
+
+size_t UARTDriver_Generic::write(const uint8_t *buffer, size_t size)
+{
+    if (!_initialised) {
+		return 0;
```
Comment (khancyr) [inline libraries/AP_HAL_ChibiOS/UARTDriver_Generic.cpp]: tabs

File: libraries/AP_IOMCU/iofirmware/iofirmware.cpp [hunk 5]
```diff
@@ -222,13 +241,25 @@ void AP_IOMCU_FW::process_io_packet()
         case CODE_READ:
         {
             num_code_read++;
-            handle_code_read();
+            if(!handle_code_read()) {
```
Comment (khancyr) [inline libraries/AP_IOMCU/iofirmware/iofirmware.cpp]: if (

File: libraries/AP_IOMCU/iofirmware/iofirmware.cpp [hunk 6]
```diff
@@ -222,13 +241,25 @@ void AP_IOMCU_FW::process_io_packet()
...
+                tx_io_packet.crc = 0;
+                tx_io_packet.crc =  crc_crc8((const uint8_t *)&tx_io_packet, tx_io_packet.get_size());
+            }
         }
         break;
         case CODE_WRITE:
         {
             num_write_pkt++;
-            handle_code_write();
+            if(!handle_code_write()) {
```
Comment (khancyr) [inline libraries/AP_IOMCU/iofirmware/iofirmware.cpp]: if (

File: Tools/ardupilotwaf/boards.py [hunk 7]
```diff
@@ -338,21 +338,18 @@ def configure_env(self, cfg, env):
             '-fno-builtin-vprintf',
             '-fno-builtin-vfprintf',
             '-fno-builtin-puts',
-            '-mcpu=cortex-m4',
             '-mno-thumb-interwork',
             '-mthumb',
-            '-mfpu=fpv4-sp-d16',
-            '-mfloat-abi=hard'
+            '--specs=nano.specs',
```
Comment (OXINARF) [inline Tools/ardupilotwaf/boards.py]: There's an extra - in there

Comment (khancyr) [reply Tools/ardupilotwaf/boards.py]: Nope those are fine!

**Comment by tridge** [comment]
I have updated this PR with the new version from @bugobliterator 
thanks Sid!

**Pull request merged**